{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2021-10-30 20:49:40 - INFO - Processing records ...\n",
      "2021-10-30 20:49:40 - INFO - Creating institutions thesaurus ...\n",
      "2021-10-30 20:49:40 - INFO - Affiliations without country detected - check file ../data/processed/debug/ignored_affiliations.txt\n",
      "2021-10-30 20:49:40 - INFO - Affiliations without country detected - check file ../data/processed/debug/ignored_affiliations.txt\n",
      "2021-10-30 20:49:40 - INFO - Thesaurus file '../data/processed/debug/institutions.txt' created.\n",
      "2021-10-30 20:49:40 - INFO - Creating keywords thesaurus ...\n",
      "2021-10-30 20:49:42 - INFO - Thesaurus file '../data/processed/debug/keywords.txt' created.\n",
      "2021-10-30 20:49:42 - INFO - Searching local references using DOI ...\n",
      "2021-10-30 20:49:48 - INFO - Searching local references using document titles ...\n",
      "2021-10-30 20:49:55 - INFO - Consolidating local references ...\n",
      "2021-10-30 20:49:55 - INFO - Computing local citations ...\n",
      "2021-10-30 20:49:55 - INFO - Computing Bradford Law Zones ...\n",
      "2021-10-30 20:49:55 - INFO - The file '../data/processed/debug/records.csv' was rewrited.\n",
      "2021-10-30 20:49:55 - INFO - Applying thesaurus ...\n",
      "2021-10-30 20:49:55 - INFO - Applying thesaurus to institutions ...\n",
      "2021-10-30 20:49:55 - INFO - Extract and cleaning institutions.\n",
      "2021-10-30 20:49:55 - INFO - Extracting institution of first author ...\n",
      "2021-10-30 20:49:55 - INFO - The file '../data/processed/debug/records.csv' was rewrited.\n",
      "2021-10-30 20:49:56 - INFO - The thesaurus was applied to institutions.\n",
      "2021-10-30 20:49:56 - INFO - Applying thesaurus to keywords ...\n",
      "2021-10-30 20:49:56 - INFO - The thesaurus was applied to keywords.\n",
      "2021-10-30 20:49:56 - INFO - Processing records finished!\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import sys\n",
    "\n",
    "currentdir = os.getcwd()\n",
    "parentdir = os.path.dirname(currentdir)\n",
    "sys.path.append(parentdir)\n",
    "\n",
    "#\n",
    "# data\n",
    "#\n",
    "from techminer import *\n",
    "\n",
    "\n",
    "# Records(directory=\"../data/processed/debug\").import_records(source=\"../data/tmp/debug-scopus.csv\", filetype=\"scopus\")\n",
    "# Records(directory=\"../data/processed/debug\").process_records()\n",
    "Records(directory=\"../data/processed/debug\").coverage()\n",
    "\n",
    "\n",
    "# wos_df = load_file(filepath=\"techminer/tests/data/savedrecs.txt\", filetype=\"wos\", datastorepath=\"./datastore\")\n",
    "# dim_df = load_file(filepath=\"techminer/tests/data/dimensions.csv\", filetype=\"dimensions\", datastorepath=\"./datastore\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "\n",
    "currentdir = os.getcwd()\n",
    "parentdir = os.path.dirname(currentdir)\n",
    "sys.path.append(parentdir)\n",
    "\n",
    "from techminer.data import *\n",
    "\n",
    "# coverage(directory_or_records=\"../data/processed/debug\")\n",
    "describe_records(directory_or_records=\"../data/processed/debug\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "\n",
    "currentdir = os.getcwd()\n",
    "parentdir = os.path.dirname(currentdir)\n",
    "sys.path.append(parentdir)\n",
    "\n",
    "#\n",
    "# features\n",
    "#\n",
    "from techminer.features import *\n",
    "\n",
    "# extract_coutries(directory_or_records=\"../data/processed/debug\", column='affiliations')\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#\n",
    "# query modules\n",
    "#\n",
    "import os\n",
    "import sys\n",
    "\n",
    "currentdir = os.getcwd()\n",
    "parentdir = os.path.dirname(currentdir)\n",
    "sys.path.append(parentdir)\n",
    "\n",
    "from techminer.query import * \n",
    "from techminer.visualization import worldmap\n",
    "\n",
    "\n",
    "\n",
    "# term_analysis(directory_or_records=\"../data/processed/debug\", column='authors')\n",
    "### count_terms_by_column(directory_or_records=\"../data/processed/debug\", sep=\"; \", column='authors')\n",
    "# time_analysis(directory_or_records=\"../data/processed/debug\")\n",
    "### count_local_citations_by_year(directory_or_records=\"../data/processed/debug\")\n",
    "### count_global_citations_by_year(directory_or_records=\"../data/processed/debug\")\n",
    "### count_documents_by_year(directory_or_records=\"../data/processed/debug\")\n",
    "### mean_global_citations_by_year(directory_or_records=\"../data/processed/debug\")\n",
    "# most_cited_documents(directory_or_records=\"../data/processed/debug\")\n",
    "# compute_impact_index(directory_or_records=\"../data/processed/debug\", column='authors')\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# ---\n",
    "# worldmap(count_documents_by_term(directory_or_records=\"../data/processed/debug\", column='countries'))\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "\n",
    "currentdir = os.getcwd()\n",
    "parentdir = os.path.dirname(currentdir)\n",
    "sys.path.append(parentdir)\n",
    "\n",
    "from techminer import Records\n",
    "\n",
    "Records(directory=)\n",
    "\n",
    "# core_sources(directory_or_records=\"../data/processed/debug\")\n",
    "# core_authors(directory_or_records=\"../data/processed/debug\")\n",
    "# lotka(core_authors(directory_or_records=\"../data/processed/debug\"), colormap='Blues', figsize=(6,3))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "\n",
    "currentdir = os.getcwd()\n",
    "parentdir = os.path.dirname(currentdir)\n",
    "sys.path.append(parentdir)\n",
    "\n",
    "\n",
    "from techminer.visualization import bradford\n",
    "\n",
    "bradford(directory_or_records=\"../data/processed/debug\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "from techminer.data.records import load_records\n",
    "\n",
    "# sorted(load_records(directory=\"../data/processed/debug\").columns)\n",
    "load_records(directory=\"../data/processed/debug\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from src.features.apply_keywords_thesaurus import apply_keywords_thesaurus\n",
    "from src.features.apply_institutions_thesaurus import apply_institutions_thesaurus\n",
    "\n",
    "# Ok!\n",
    "apply_keywords_thesaurus(datastoredir=\"../data/processed/debug\")\n",
    "apply_institutions_thesaurus(datastoredir=\"../data/processed/debug\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "\n",
    "currentdir = os.getcwd()\n",
    "parentdir = os.path.dirname(currentdir)\n",
    "sys.path.append(parentdir)\n",
    "\n",
    "from techminer.analysis import Bradford\n",
    "\n",
    "bradford = Bradford(datastoredir=\"../data/processed/debug\")\n",
    "\n",
    "bradford.plot(max_items=8, cmap='Blues', figsize=(8, 5), fontsize=9)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "bradford.sources_"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "\n",
    "currentdir = os.getcwd()\n",
    "parentdir = os.path.dirname(currentdir)\n",
    "sys.path.append(parentdir)\n",
    "\n",
    "from src.analysis import TopDocuments\n",
    "\n",
    "TopDocuments(datastoredir=\"../data/processed/debug\").detailed_"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from techminer.analysis.top_documents import TopDocuments\n",
    "\n",
    "TopDocuments(datastorepath=\"./datastore\").print()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from techminer.analysis.core_authors import CoreAuthors\n",
    "\n",
    "CoreAuthors(datastorepath=\"./datastore\").table_"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from techminer.analysis.core_authors import CoreAuthors\n",
    "\n",
    "CoreAuthors(datastorepath=\"./datastore\").plot(figsize=(8, 4))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from techminer.analysis.core_sources import CoreSources\n",
    "\n",
    "CoreSources(datastorepath=\"./datastore\").table_"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from techminer.analysis.coverage import Coverage\n",
    "\n",
    "Coverage(datastorepath=\"./datastore\").table_"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from techminer.analysis.worldmap import WorldMap\n",
    "\n",
    "WorldMap(datastorepath=\"./datastore\").plot(figsize=(9, 5), colormap='Blues')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>authors</th>\n",
       "      <th>authors_id</th>\n",
       "      <th>document_title</th>\n",
       "      <th>pub_year</th>\n",
       "      <th>publication_name</th>\n",
       "      <th>global_citations</th>\n",
       "      <th>doi</th>\n",
       "      <th>affiliations</th>\n",
       "      <th>abstract</th>\n",
       "      <th>author_keywords</th>\n",
       "      <th>...</th>\n",
       "      <th>global_references_count</th>\n",
       "      <th>historiograph_id</th>\n",
       "      <th>local_references</th>\n",
       "      <th>local_citations</th>\n",
       "      <th>bradford_law_zone</th>\n",
       "      <th>institutions</th>\n",
       "      <th>institution_1st_author</th>\n",
       "      <th>index_keywords_cl</th>\n",
       "      <th>keywords</th>\n",
       "      <th>record_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Krishnamoorthy A; Mishra A; Kamal D; Hong S; N...</td>\n",
       "      <td>Krishnamoorthy A/57198292319; Mishra A/5719779...</td>\n",
       "      <td>EZFF: Python library for multi-objective param...</td>\n",
       "      <td>2021</td>\n",
       "      <td>SOFTWAREX</td>\n",
       "      <td>0</td>\n",
       "      <td>10.1016/J.SOFTX.2021.100663</td>\n",
       "      <td>Collaboratory of Advanced Computing and Simula...</td>\n",
       "      <td>parameterization of interatomic forcefields is...</td>\n",
       "      <td>genetic algorithm; global optimization; intera...</td>\n",
       "      <td>...</td>\n",
       "      <td>49</td>\n",
       "      <td>2021-0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>University of Southern California USA; Georgia...</td>\n",
       "      <td>University of Southern California USA</td>\n",
       "      <td>engines; genetic algorithm; global optimizatio...</td>\n",
       "      <td>functional forms; genetic algorithms; global ...</td>\n",
       "      <td>2021-0001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Wagner S; Thierbach K; Zerjatke T; Glauche I; ...</td>\n",
       "      <td>Wagner S/57221091683; Thierbach K/55860831500;...</td>\n",
       "      <td>TraCurate: Efficiently curating cell tracks</td>\n",
       "      <td>2021</td>\n",
       "      <td>SOFTWAREX</td>\n",
       "      <td>0</td>\n",
       "      <td>10.1016/J.SOFTX.2021.100656</td>\n",
       "      <td>Institute for Medical Informatics and Biometry...</td>\n",
       "      <td>tracurate is an open-source software tool to c...</td>\n",
       "      <td>image analysis; lineage tracing; single cell t...</td>\n",
       "      <td>...</td>\n",
       "      <td>36</td>\n",
       "      <td>2021-1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>Institute for Medical Informatics and Biometry...</td>\n",
       "      <td>Institute for Medical Informatics and Biometry...</td>\n",
       "      <td>cells; computer operating systems; open-source...</td>\n",
       "      <td>automated tracking; cell tracking; computer o...</td>\n",
       "      <td>2021-0023</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Arslan ŞŞ</td>\n",
       "      <td>Arslan ŞŞ/35955672100</td>\n",
       "      <td>Founsure 1.0: An erasure code library with eff...</td>\n",
       "      <td>2021</td>\n",
       "      <td>SOFTWAREX</td>\n",
       "      <td>0</td>\n",
       "      <td>10.1016/J.SOFTX.2021.100662</td>\n",
       "      <td>MEF University, Maslak, Istanbul, Turkey</td>\n",
       "      <td>founsure is an open-source software library th...</td>\n",
       "      <td>distributed storage; erasure coding; fountain ...</td>\n",
       "      <td>...</td>\n",
       "      <td>23</td>\n",
       "      <td>2021-2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>Mef University TUR</td>\n",
       "      <td>Mef University TUR</td>\n",
       "      <td>bandwidth; codes (symbols); data handling; dig...</td>\n",
       "      <td>codes (symbols); compiler optimizations; comp...</td>\n",
       "      <td>2021-0018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Wacha A; Beke-Somfai T</td>\n",
       "      <td>Wacha A/37040010900; Beke-Somfai T/35323780200</td>\n",
       "      <td>PmlBeta: A PyMOL extension for building β-amin...</td>\n",
       "      <td>2021</td>\n",
       "      <td>SOFTWAREX</td>\n",
       "      <td>0</td>\n",
       "      <td>10.1016/J.SOFTX.2020.100654</td>\n",
       "      <td>Institute of Materials and Environmental Chemi...</td>\n",
       "      <td>β-peptides are peptide-mimetics with wide biom...</td>\n",
       "      <td>foldamers; pymol; β-peptides</td>\n",
       "      <td>...</td>\n",
       "      <td>30</td>\n",
       "      <td>2021-3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>Institute of Materials and Environmental Chemi...</td>\n",
       "      <td>Institute of Materials and Environmental Chemi...</td>\n",
       "      <td>amino acids; bioinformatics; molecular graphic...</td>\n",
       "      <td>bioinformatics; biotechnological applications...</td>\n",
       "      <td>2021-0017</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Bach E</td>\n",
       "      <td>Bach E/57200651611</td>\n",
       "      <td>parasweep: A template-based utility for genera...</td>\n",
       "      <td>2021</td>\n",
       "      <td>SOFTWAREX</td>\n",
       "      <td>0</td>\n",
       "      <td>10.1016/J.SOFTX.2020.100631</td>\n",
       "      <td>Department of Atmospheric and Oceanic Science ...</td>\n",
       "      <td>we introduce parasweep, a free and open-source...</td>\n",
       "      <td>distributed computing; parallel computing; par...</td>\n",
       "      <td>...</td>\n",
       "      <td>19</td>\n",
       "      <td>2021-4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>University of Maryland USA</td>\n",
       "      <td>University of Maryland USA</td>\n",
       "      <td>code changes; computational model; configurati...</td>\n",
       "      <td>computational model; configuration files; hig...</td>\n",
       "      <td>2021-0016</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 27 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                             authors  \\\n",
       "0  Krishnamoorthy A; Mishra A; Kamal D; Hong S; N...   \n",
       "1  Wagner S; Thierbach K; Zerjatke T; Glauche I; ...   \n",
       "2                                          Arslan ŞŞ   \n",
       "3                             Wacha A; Beke-Somfai T   \n",
       "4                                             Bach E   \n",
       "\n",
       "                                          authors_id  \\\n",
       "0  Krishnamoorthy A/57198292319; Mishra A/5719779...   \n",
       "1  Wagner S/57221091683; Thierbach K/55860831500;...   \n",
       "2                              Arslan ŞŞ/35955672100   \n",
       "3     Wacha A/37040010900; Beke-Somfai T/35323780200   \n",
       "4                                 Bach E/57200651611   \n",
       "\n",
       "                                      document_title  pub_year  \\\n",
       "0  EZFF: Python library for multi-objective param...      2021   \n",
       "1        TraCurate: Efficiently curating cell tracks      2021   \n",
       "2  Founsure 1.0: An erasure code library with eff...      2021   \n",
       "3  PmlBeta: A PyMOL extension for building β-amin...      2021   \n",
       "4  parasweep: A template-based utility for genera...      2021   \n",
       "\n",
       "  publication_name  global_citations                          doi  \\\n",
       "0        SOFTWAREX                 0  10.1016/J.SOFTX.2021.100663   \n",
       "1        SOFTWAREX                 0  10.1016/J.SOFTX.2021.100656   \n",
       "2        SOFTWAREX                 0  10.1016/J.SOFTX.2021.100662   \n",
       "3        SOFTWAREX                 0  10.1016/J.SOFTX.2020.100654   \n",
       "4        SOFTWAREX                 0  10.1016/J.SOFTX.2020.100631   \n",
       "\n",
       "                                        affiliations  \\\n",
       "0  Collaboratory of Advanced Computing and Simula...   \n",
       "1  Institute for Medical Informatics and Biometry...   \n",
       "2           MEF University, Maslak, Istanbul, Turkey   \n",
       "3  Institute of Materials and Environmental Chemi...   \n",
       "4  Department of Atmospheric and Oceanic Science ...   \n",
       "\n",
       "                                            abstract  \\\n",
       "0  parameterization of interatomic forcefields is...   \n",
       "1  tracurate is an open-source software tool to c...   \n",
       "2  founsure is an open-source software library th...   \n",
       "3  β-peptides are peptide-mimetics with wide biom...   \n",
       "4  we introduce parasweep, a free and open-source...   \n",
       "\n",
       "                                     author_keywords  ...  \\\n",
       "0  genetic algorithm; global optimization; intera...  ...   \n",
       "1  image analysis; lineage tracing; single cell t...  ...   \n",
       "2  distributed storage; erasure coding; fountain ...  ...   \n",
       "3                       foldamers; pymol; β-peptides  ...   \n",
       "4  distributed computing; parallel computing; par...  ...   \n",
       "\n",
       "  global_references_count historiograph_id local_references  local_citations  \\\n",
       "0                      49           2021-0              NaN                0   \n",
       "1                      36           2021-1              NaN                0   \n",
       "2                      23           2021-2              NaN                0   \n",
       "3                      30           2021-3              NaN                0   \n",
       "4                      19           2021-4              NaN                0   \n",
       "\n",
       "   bradford_law_zone                                       institutions  \\\n",
       "0                  3  University of Southern California USA; Georgia...   \n",
       "1                  3  Institute for Medical Informatics and Biometry...   \n",
       "2                  3                                 Mef University TUR   \n",
       "3                  3  Institute of Materials and Environmental Chemi...   \n",
       "4                  3                         University of Maryland USA   \n",
       "\n",
       "                              institution_1st_author  \\\n",
       "0              University of Southern California USA   \n",
       "1  Institute for Medical Informatics and Biometry...   \n",
       "2                                 Mef University TUR   \n",
       "3  Institute of Materials and Environmental Chemi...   \n",
       "4                         University of Maryland USA   \n",
       "\n",
       "                                   index_keywords_cl  \\\n",
       "0  engines; genetic algorithm; global optimizatio...   \n",
       "1  cells; computer operating systems; open-source...   \n",
       "2  bandwidth; codes (symbols); data handling; dig...   \n",
       "3  amino acids; bioinformatics; molecular graphic...   \n",
       "4  code changes; computational model; configurati...   \n",
       "\n",
       "                                            keywords  record_id  \n",
       "0   functional forms; genetic algorithms; global ...  2021-0001  \n",
       "1   automated tracking; cell tracking; computer o...  2021-0023  \n",
       "2   codes (symbols); compiler optimizations; comp...  2021-0018  \n",
       "3   bioinformatics; biotechnological applications...  2021-0017  \n",
       "4   computational model; configuration files; hig...  2021-0016  \n",
       "\n",
       "[5 rows x 27 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "pd.read_csv(\"../data/processed/debug/records.csv\").head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "949777d72b0d2535278d3dc13498b2535136f6dfe0678499012e853ee9abcab1"
  },
  "kernelspec": {
   "display_name": "Python 3.7.12 64-bit",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
